packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -490,196 +501,183 @@ export function getValidatorApi({
+          builderPayloadValue: `${builderPayloadValue}`,
+          builderConsensusValue: `${builderConsensusValue}`,
+          builderBlockValue: `${builderBlockValue}`,
         });
-      } else if (fullBlock && blindedBlock) {
-        switch (builderSelection) {
-          case routes.validator.BuilderSelection.MaxProfit: {
-            if (
-              // explicitly handle the two special values mentioned in spec for builder preferred / engine preferred
-              builderBoostFactor !== MAX_BUILDER_BOOST_FACTOR &&
-              (builderBoostFactor === BigInt(0) ||
-                blockValueEngine >= (blockValueBuilder * builderBoostFactor) / BigInt(100))
-            ) {
-              executionPayloadSource = ProducedBlockSource.engine;
-            } else {
-              executionPayloadSource = ProducedBlockSource.builder;
-            }
-            break;
-          }
 
-          case routes.validator.BuilderSelection.ExecutionOnly: {
-            executionPayloadSource = ProducedBlockSource.engine;
-            break;
-          }
+        return {...builder.value, executionPayloadBlinded: true, executionPayloadSource: ProducedBlockSource.builder};
+      }
 
-          // For everything else just select the builder
-          default: {
-            executionPayloadSource = ProducedBlockSource.builder;
-          }
+      if (engine.status === "fulfilled" && builder.status !== "fulfilled") {
+        const engineBlock = engine.value;
+        const enginePayloadValue = engineBlock.executionPayloadValue ?? BigInt(0);
+        const engineConsensusValue = engineBlock.consensusBlockValue ?? BigInt(0);
+        const engineBlockValue = enginePayloadValue + engineConsensusValue;
+
+        logger.info("Selected engine block: no builder block produced", {
+          durationMs: engine.durationMs,
+          slot,
+          enginePayloadValue: `${enginePayloadValue}`,
+          engineConsensusValue: `${engineConsensusValue}`,
+          engineBlockValue: `${engineBlockValue}`,
+        });
+
+        return {...engine.value, executionPayloadBlinded: false, executionPayloadSource: ProducedBlockSource.engine};
+      }
+
+      if (engine.status === "fulfilled" && builder.status === "fulfilled") {
+        const engineBlock = engine.value;
+        const enginePayloadValue = engineBlock.executionPayloadValue ?? BigInt(0);
+        const engineConsensusValue = engineBlock.consensusBlockValue ?? BigInt(0);
+        const engineBlockValue = enginePayloadValue + engineConsensusValue;
+
+        const builderBlock = builder.value;
+        const builderPayloadValue = builderBlock.executionPayloadValue ?? BigInt(0);
+        const builderConsensusValue = builderBlock.consensusBlockValue ?? BigInt(0);
+        const builderBlockValue = builderPayloadValue + builderConsensusValue;
+
+        if (engineBlock.shouldOverrideBuilder) {
+          logger.info("Selected engine block as censorship suspected in builder blocks", {
```

Comment:
Previously, if `shouldOverrideBuilder` was returned from EL this log would be trigger independent of whether or not the builder block resolved.

I think if `shouldOverrideBuilder` is set by EL it should have prio 1, and we need to log this no matter if builder block succeeds or not.

Reply:
+1

Reply:
We can override builder block only when builder produce a block. If builder failed to produce a block then no matter what value is given by engine, we can't override a non-existing block.

Reply:
The outcome (choosing engine block) is the same but not  why it was chosen. Communicating to the user that builders are suspected to censor transactions is very important and should have prio 1.

Reply:
We are communicating in much clearer way now. If builder block failed to produce, we don't go and tell user that we selected engine block because of censorship. We tell user that because builder is failed, so we choose engine. 

The question of censorship or additional logic only come into effect once both builder and engine produced the blocks.

Reply:
Our logic is wrong then, ideally we shouldn't even check if builder resolved or not and even return early as noted in https://github.com/ChainSafe/lodestar/pull/6241#discussion_r1462947877.

My reasoning about this is purely from a network health perspecitve and what is more important to communicate to the user.

Reply:
It's exactly the same behavior we have earlier. We wait till **cut-off** time for both and then move further steps in execution. 

If call to builder and engine was sequential then I would agree to network health point, but we do both calls in parallel, so regardless we wait for the cutoff time or not request is already sent out. 

And additionally such behaviors should be designed on heuristics not the edge cases assumptions. I don't think there is quite difference in builder and engine performance that we should be thinking more into it even before cut-off time. In future if we have more data, and we observe the a major time difference between both then we would think to work on earlier exit logic based on `shouldOverrideBuilder`